ext/session/session.c

```diff
@@ -1600,6 +1618,12 @@ PHPAPI zend_result php_session_start(void) /* {{{ */
 		}
 		return FAILURE;
 	}
+
+	/* Should these be set here, or in session_initialize? */
+	php_session_cleanup_filename();
+	PS(session_started_filename) = estrdup(zend_get_executed_filename());
```

Comment:
Can't we just use `zend_get_executed_filename_ex` to avoid the string duplication?

Reply:
This diff would work:

```diff
diff --git a/ext/session/php_session.h b/ext/session/php_session.h
index 2f26cd16b5..341aac5716 100644
--- a/ext/session/php_session.h
+++ b/ext/session/php_session.h
@@ -154,7 +154,7 @@ typedef struct _php_ps_globals {
 	const ps_module *default_mod;
 	void *mod_data;
 	php_session_status session_status;
-	char *session_started_filename;
+	zend_string *session_started_filename;
 	uint32_t session_started_lineno;
 	zend_long gc_probability;
 	zend_long gc_divisor;
diff --git a/ext/session/session.c b/ext/session/session.c
index f853da4b65..aaaaf98b38 100644
--- a/ext/session/session.c
+++ b/ext/session/session.c
@@ -126,7 +126,7 @@ static inline void php_rinit_session_globals(void) /* {{{ */
 static inline void php_session_cleanup_filename(void) /* {{{ */
 {
 	if (PS(session_started_filename)) {
-		efree(PS(session_started_filename));
+		zend_string_release_ex(PS(session_started_filename), 0);
 		PS(session_started_filename) = NULL;
 		PS(session_started_lineno) = 0;
 	}
@@ -1505,7 +1505,7 @@ PHPAPI zend_result php_session_start(void) /* {{{ */
 	switch (PS(session_status)) {
 		case php_session_active:
 			if (PS(session_started_filename)) {
-				php_error(E_NOTICE, "Ignoring session_start() because a session has already been started (started from %s on line %u)", PS(session_started_filename), PS(session_started_lineno));
+				php_error(E_NOTICE, "Ignoring session_start() because a session has already been started (started from %s on line %u)", ZSTR_VAL(PS(session_started_filename)), PS(session_started_lineno));
 			} else {
 				php_error(E_NOTICE, "Ignoring session_start() because a session has already been started");
 			}
@@ -1621,8 +1621,11 @@ PHPAPI zend_result php_session_start(void) /* {{{ */
 
 	/* Should these be set here, or in session_initialize? */
 	php_session_cleanup_filename();
-	PS(session_started_filename) = estrdup(zend_get_executed_filename());
-	PS(session_started_lineno) = zend_get_executed_lineno();
+	zend_string *session_started_filename = zend_get_executed_filename_ex();
+	if (session_started_filename != NULL) {
+		PS(session_started_filename) = zend_string_copy(session_started_filename);
+		PS(session_started_lineno) = zend_get_executed_lineno();
+	}
 
 	return SUCCESS;
 }
@@ -2538,7 +2541,7 @@ PHP_FUNCTION(session_start)
 
 	if (PS(session_status) == php_session_active) {
 		if (PS(session_started_filename)) {
-			php_error_docref(NULL, E_NOTICE, "Ignoring session_start() because a session is already active (started from %s on line %d)", PS(session_started_filename), PS(session_started_lineno));
+			php_error_docref(NULL, E_NOTICE, "Ignoring session_start() because a session is already active (started from %s on line %d)", ZSTR_VAL(PS(session_started_filename)), PS(session_started_lineno));
 		} else {
 			php_error_docref(NULL, E_NOTICE, "Ignoring session_start() because a session is already active");
 		}
```

...but it breaks `ext/session/tests/session_start_variation9.diff`:

```
     *** Testing session_start() : variation ***
     string(%d) "%s"
     
004+ Notice: session_start(): Ignoring session_start() because a session is already active in /home/calvin/src/php-src/ext/session/tests/session_start_variation9.php on line 8
004- Notice: session_start(): Ignoring session_start() because a session is already active (started from %s on line %d) in %s on line %d
     bool(true)
     string(%d) "%s"
     bool(true)
--
```